src/raftstore/store/snap/builder.rs

```diff
@@ -0,0 +1,465 @@
+use std::fs::{self, File, OpenOptions};
+use std::io::{self, prelude::*};
+use std::sync::Arc;
+use std::time::Instant;
+
+use crc::crc32::{self, Digest, Hasher32};
+use kvproto::metapb::Region;
+use rocksdb::{DBCompressionType, EnvOptions, SstFileWriter};
+
+use raftstore::store::engine::{Iterable, Snapshot as DbSnapshot};
+use raftstore::store::keys::{enc_end_key, enc_start_key};
+use raftstore::store::metrics::*;
+use raftstore::store::snap::*;
+use raftstore::Result as RaftStoreResult;
+use storage::CfName;
+use util::codec::bytes::BytesEncoder;
+use util::file::{calc_crc32, delete_file_if_exist, get_file_size};
+use util::rocksdb::get_fastest_supported_compression_type;
+use util::time::duration_to_sec;
+use util::Either;
+
+const TMP_FILE_SUFFIX: &str = ".tmp";
+
+struct CfFile {
+    cf: CfName,
+    tmp_cf_path: PathBuf,
+    tmp_cf_file: Either<File, SstFileWriter>,
+    digest: Digest,
+    written_bytes: u64,
+}
+
+impl CfFile {
+    fn new(cf: CfName, tmp_cf_path: PathBuf, tmp_cf_file: Either<File, SstFileWriter>) -> Self {
+        CfFile {
+            cf,
+            tmp_cf_path,
+            tmp_cf_file,
+            digest: Digest::new(crc32::IEEE),
+            written_bytes: 0,
+        }
+    }
+
+    fn scan_cf(
+        &mut self,
+        db_snap: &DbSnapshot,
+        region_id: u64,
+        start_key: &[u8],
+        end_key: &[u8],
+        io_limiter: Option<&IOLimiter>,
+        snap_key: SnapKey,
+        snap_stale_notifier: &SnapStaleNotifier,
+    ) -> RaftStoreResult<usize> {
+        let (mut cf_key_count, mut cf_size) = (0, 0);
+        match self.tmp_cf_file {
+            Either::Left(ref mut file) => {
+                db_snap.scan_cf(self.cf, start_key, end_key, false, |key, value| {
+                    cf_key_count += 1;
+                    cf_size += key.len() + value.len();
+                    file.encode_compact_bytes(key)?;
+                    file.encode_compact_bytes(value)?;
+                    Ok(!stale_for_generate(snap_key, snap_stale_notifier))
+                })?;
+                // use an empty byte array to indicate that cf reaches an end.
+                file.encode_compact_bytes(b"")?;
+            }
+            Either::Right(ref mut writer) => {
+                let mut bytes = 0;
+                let base = io_limiter.map_or(0, |l| l.get_max_bytes_per_time());
+                db_snap.scan_cf(self.cf, &start_key, &end_key, false, |key, value| {
+                    let l = key.len() + value.len();
+                    cf_key_count += 1;
+                    cf_size += l;
+                    if let Some(ref limiter) = io_limiter {
+                        if bytes >= base {
+                            bytes = 0;
+                            limiter.request(base);
+                        }
+                        bytes += l as i64;
+                    }
+                    writer.put(key, value)?;
+                    Ok(!stale_for_generate(snap_key, snap_stale_notifier))
```

Comment:
Have you test the performance of function `stale_for_generate`, because it will be called for each key/value pair.

Reply:
I add a bench test for this.